frontend/src/lib/WidgetStateManager.ts

```diff
@@ -595,37 +603,63 @@ export class WidgetStateManager {
   }
 
   /**
-   * Called by FormSubmitButton on creation. Increment submitButtonCount for
-   * the given form, and update FormsData.
+   * Called by FormSubmitButton on creation. Add the SubmitButtonProto for
+   * the given form and update FormsData.
    */
-  public incrementSubmitButtonCount(formId: string): void {
-    this.setSubmitButtonCount(
-      formId,
-      WidgetStateManager.getSubmitButtonCount(this.formsData, formId) + 1
-    )
+  public addSubmitButton(
+    formId: string,
+    submitButtonProto: SubmitButtonProto
+  ): void {
+    const submitButtons = this.formsData.submitButtons.get(formId)
+    if (submitButtons === undefined) {
+      this.setSubmitButtons(formId, [submitButtonProto])
+    } else {
+      const copySubmitButtons = Object.assign([], submitButtons)
+      copySubmitButtons.push(submitButtonProto)
+      this.setSubmitButtons(formId, copySubmitButtons)
+    }
   }
 
   /**
-   * Called by FormSubmitButton on creation. Decrement submitButtonCount for
-   * the given form, and update FormsData.
+   * Called by FormSubmitButton on creation. Remove the SubmitButtonProto for
+   * the given form and update FormsData.
    */
-  public decrementSubmitButtonCount(formId: string): void {
-    this.setSubmitButtonCount(
-      formId,
-      WidgetStateManager.getSubmitButtonCount(this.formsData, formId) - 1
-    )
+  public removeSubmitButton(
+    formId: string,
+    submitButtonProto: SubmitButtonProto
+  ): void {
+    const submitButtons = this.formsData.submitButtons.get(formId)
+    if (submitButtons !== undefined) {
+      const copySubmitButtons = Object.assign([], submitButtons)
+      const index = copySubmitButtons.indexOf(submitButtonProto, 0)
+      if (index > -1) {
+        copySubmitButtons.splice(index, 1)
+      }
+      this.setSubmitButtons(formId, copySubmitButtons)
+    }
   }
 
-  private setSubmitButtonCount(formId: string, count: number): void {
-    if (count < 0) {
-      throw new Error(`Bad submitButtonCount value ${count} (must be >= 0)`)
+  private setSubmitButtons(
+    formId: string,
+    submitButtons: Array<SubmitButtonProto>
+  ): void {
+    if (submitButtons.length < 0) {
+      throw new Error(
+        `Bad submitButtons value ${submitButtons.length} (must be >= 0)`
+      )
     }
 
     this.updateFormsData(draft => {
-      draft.submitButtonCount.set(formId, count)
+      draft.submitButtons.set(formId, submitButtons)
     })
   }
 
+  public getSubmitButton(
```

Comment:
Should you be able to call this method in `utils.ts` with a `WidgetStateManager`? Or should we just simply have have each Widget's`widgetMgr` call `submitForm` themselves? 

Seemed like a duplication of code so I tried to refactor it here.